File: 1744-number-of-ways-to-form-a-target-string-given-a-dictionary/1744-number-of-ways-to-form-a-target-string-given-a-dictionary.py
class Solution:
    def numWays(self, words: List[str], target: str) -> int:
        wordLen = len(words[0])
        letterMap = [[0 for _ in range(26)] for _ in range(wordLen)]
        mod = 10 ** 9 + 7
        
        for word in words:
            for i, letter in enumerate(word):
                letterMap[i][ord(letter) - ord('a')] += 1
        
        memo = {}
        def dp(i, k):
            if (i, k) in memo: 
                return memo[(i, k)]

            if i >= len(target): 
                return 1

            if k >= wordLen: 
                return 0
            
            ans = letterMap[k][ord(target[i]) - ord('a')] * dp(i + 1, k + 1) + dp(i, k + 1)
            
            memo[(i, k)] = ans

            return ans % mod
        
        return dp(0, 0)

File: 1744-number-of-ways-to-form-a-target-string-given-a-dictionary/1744-number-of-ways-to-form-a-target-string-given-a-dictionary.cpp
class Solution {
public:
    int numWays(vector<string>& words, string target) {
        int mod = 1e9 + 7;
        unordered_map<int, unordered_map<char, int>> letterFreq;
        int maxJ = 0;
        for (int i = 0; i < words.size(); i++) {
            for (int j = 0; j < words[i].size(); j++) {
                letterFreq[j][words[i][j]]++;
                maxJ = max(j, maxJ);
            }
        }

        vector<vector<long>> memo(maxJ + 1, vector<long>(target.size() + 1, -1));

        function<long(int, int)> dp = [&] (int i, int j) {
            if (j >= target.size()) {
                return 1l;
            }
            if (i > maxJ) {
                return 0l;
            }
            if (memo[i][j] != -1) return memo[i][j];
            long ans = dp(i + 1, j) % mod;

            if (letterFreq[i][target[j]] > 0) {
                ans += (letterFreq[i][target[j]] * dp(i + 1, j + 1)) % mod;
            }
            memo[i][j] = ans % mod;
            return memo[i][j];
        };

        return dp(0, 0);
    }
};